Add new `Lint/RedundantLiteralConversion` cop

dvandersluis opened the pull request:
This new cop looks for instances where conversation methods calls (`to_s`, `to_sym`, `to_i`, `to_f`, `to_r`, `to_c`, `to_a`, `to_h` and `to_set`) are redundant -- when chained to a literal of the conversion type, or to another conversion call of the same type (ie. `to_s.to_s`, etc.).

-----------------

Before submitting the PR make sure the following are checked:

* [x] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences.
* [x] Wrote [good commit messages][1].
* [ ] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Ran `bundle exec rake default`. It executes all tests and runs RuboCop on its own code.
* [x] Added an entry (file) to the [changelog folder](https://github.com/rubocop/rubocop/blob/master/changelog/) named `{change_type}_{change_description}.md` if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format) for details.

[1]: https://chris.beams.io/posts/git-commit/

Earlopain commented on lib/rubocop/cop/style/redundant_to_s.rb:
```diff
@@ -0,0 +1,101 @@
+      # heredoc, or when chained with itself.
+      #
+      # @safety
+      #   Autocorrect is unsafe because the cop assumes that `to_s` called on an object returns
```
Is this something to realistically consider?  `to_s` returning a string seems like a fundamental requirement.

dvandersluis replied:
I was trying to hedge any bets, but I’m ok with making it safe too. I’d agree that if `to_s` is doing something wonky there’s bigger problems.

Earlopain replied:
I think many more cops would need be unsafe if you consider each and every method to be overwritable. It's technically true but I'd just drop this, yeah.

koic replied:
Yeah, when considering changes that touch upon the core of the language specification, the number of cops that can truly be deemed "safe" decreases. So this cop should be safe.

dvandersluis replied:
Sounds good, will do.

dvandersluis commented:
Makes sense, I'll look at extending it to other literals tomorrow. Thanks for the feedback!

bbatsov commented:
>  This would help differentiate its purpose from the Lint/RedundantStringCoercion cop.

I also wanted to point out we should differentiate this cop from the existing one (or expand the existing one). I also wanted to suggest the name `RedundantStringConversion` as a more generic one. (and perhaps easier to understand at a glance)

dvandersluis commented:
I've updated the cop as follows:
* It is now called `Lint/RedundantLiteralConversion`
* Removed `SafeAutoCorrect: false`
* It now covers `to_s`, `to_sym`, `to_i`, `to_f`, `to_r`, `to_c`, `to_a`, `to_h` and `to_set`, and all the literals/constructors for each type. eg. `'string'.to_s`, `String('string').to_s`, `String.new('string').to_s`, etc. A full list is described in the cop documentation.

Technically, `String('string')`, etc. aren't literals per se, so if desired, I thought `Lint/RedundantSelfConversion` might be an alternative name (if there's no concerns of confusion with `self`).

> Technically, there would be some overlap with the existing `Lint/SymbolConversion` and cases like :sym.to_sym, but it should not be a problem.

This is now true, both cops will register an offense on `:sym.to_sym`, and both make the same autocorrection so there is no clobbering issue. Would it be prefered to remove the check from `Lint/SymbolConversion` and keep it for converting strings to symbols?

bbatsov commented:
> Technically, String('string'), etc. aren't literals per se, so if desired, I thought Lint/RedundantSelfConversion might be an alternative name (if there's no concerns of confusion with self).

Or maybe just `RedundantConversion`, as I think the `Self` part will be misleading due to `self`. 

> This is now true, both cops will register an offense on :sym.to_sym, and both make the same autocorrection so there is no clobbering issue. Would it be prefered to remove the check from Lint/SymbolConversion and keep it for converting strings to symbols?

I'm on the fence about this, but leaning towards doing this check only in the new cop. 

I was also thinking we can extend the cop a bit to cover other common methods with well-known return types. (e.g. something like `inspect`) Probably this would make the cop more useful as the things it catches right now probably happen quite rarely in practice. 

@koic @Earlopain What do you think?

koic commented:
`RedundantConversion` might be a bit abstract. I haven’t thought of a better name yet, but it would be ideal to have a slightly more specific name. At the very least, since it now targets more than just literals (e.g., `String.new`), I agree with a broader name than `RedundantLiteralConversion`.

> Would it be prefered to remove the check from Lint/SymbolConversion and keep it for converting strings to symbols?

Yeah, that seems reasonable. It might be a good idea to remove the functionality for `:sym.to_sym` from `Lint/SymbolConversion`. However, since it’s not a conflicting issue, it can be addressed later as well :-)

koic commented on spec/rubocop/cop/lint/redundant_literal_conversion_spec.rb:
```diff
@@ -0,0 +1,266 @@
+    RUBY
+  end
+
+  context '`to_s`' do
```
These cases should be in `describe`, not `context`.

```suggestion
  describe '`to_s`' do
```

dvandersluis replied:
Thank you, updated.

Earlopain commented:
> I was also thinking we can extend the cop a bit to cover other common methods with well-known return types

This does make sense to me, although I can't quickly come up with more other than `inspect`. You do start risking people overwriting these, with `inspect` I wouldn't care but maybe for other methods it would be more of a concern.

I don't think code like `1.to_i` is very common, so making it a bit more useful is a good idea. Although, even for `inspect.to_s` I would doubt that many would write such code so I don't think it will see much practical use even with that.

--------------

Maybe `Lint/RedundantTypeConversion` could be a good, to keep close to the original name? That one sounded fine to me.

dvandersluis commented:
Sorry guys I thought I had addressed your comments and realized I hadn't yet.

I pushed an update with the following changes:
1. The name was changed to `RedundantTypeConversion` as per @Earlopain's suggestion.
2. The cop can now detect `inspect.to_s` and correct it as well. I am not sure what other methods would make sense to extend this to. That being said, I set it up so that it'd be easy to extend with other methods in the future as they arise.

> I don't think code like 1.to_i is very common, so making it a bit more useful is a good idea. Although, even for inspect.to_s I would doubt that many would write such code so I don't think it will see much practical use even with that.

I started building this cop because I had found a scenario where multiple cops' correction resulted in something like `foo.to_s.to_s`. It's not I can't remember anymore what cop it was unfortunately, but I thought it'd be good to build off of that.

formigarafa commented:
how about the case `Integer("1_number", exception: false).to_i  # => 0`

dvandersluis commented:
@formigarafa please see #13890.
